builder/amazon/common/step_create_ssm_tunnel.go

```diff
@@ -0,0 +1,135 @@
+	params := map[string][]*string{
+		"portNumber":      []*string{aws.String(dst)},
+		"localPortNumber": []*string{aws.String(src)},
+	}
+	l.Close()
+
+	instance, ok := state.Get("instance").(*ec2.Instance)
+	if !ok {
+		err := fmt.Errorf("error encountered in obtaining target instance id for SSM tunnel")
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	s.instanceId = aws.StringValue(instance.InstanceId)
+	ssmconn := ssm.New(s.AWSSession)
+	input := ssm.StartSessionInput{
+		DocumentName: aws.String("AWS-StartPortForwardingSession"),
+		Parameters:   params,
+		Target:       aws.String(s.instanceId),
+	}
+
+	ui.Message(fmt.Sprintf("Starting PortForwarding session to instance %q on local port %q to remote port %q", s.instanceId, src, dst))
+	var output *ssm.StartSessionOutput
+	err = retry.Config{
+		ShouldRetry: func(err error) bool { return isAWSErr(err, "TargetNotConnected", "") },
+		RetryDelay:  (&retry.Backoff{InitialBackoff: 200 * time.Millisecond, MaxBackoff: 60 * time.Second, Multiplier: 2}).Linear,
+	}.Run(ctx, func(ctx context.Context) error {
+		output, err = ssmconn.StartSessionWithContext(ctx, &input)
+		return err
+	})
+
+	if err != nil {
+		err = fmt.Errorf("error encountered in starting session for instance %q: %s", s.instanceId, err)
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+	s.ssmSession = output
+
+	// AWS session-manager-plugin requires a valid session be passed in JSON
+	sessionDetails, err := json.Marshal(s.ssmSession)
+	if err != nil {
+		ui.Error(err.Error())
+		state.Put("error encountered in reading session details", err)
+		return multistep.ActionHalt
+	}
+
+	sessionParameters, err := json.Marshal(input)
+	if err != nil {
+		ui.Error(err.Error())
+		state.Put("error encountered in reading session parameter details", err)
+		return multistep.ActionHalt
+	}
+
+	driver := SSMDriver{Ui: ui}
+	// sessionDetails, region, "StartSession", profile, paramJson, endpoint
+	region := aws.StringValue(s.AWSSession.Config.Region)
+	// how to best get Profile name
+	if err := driver.StartSession(string(sessionDetails), region, "", string(sessionParameters), ssmconn.Endpoint); err != nil {
+		err = fmt.Errorf("error encountered in establishing a tunnel with the session-manager-plugin: %s", err)
+		ui.Error(err.Error())
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	ui.Message(fmt.Sprintf("PortForwarding session to instance %q established!", s.instanceId))
+	state.Put("sessionPort", l.Port)
+
+	return multistep.ActionContinue
+}
+
+func (s *StepCreateSSMTunnel) Cleanup(state multistep.StateBag) {
+	if s.ssmSession == nil {
+		return
+	}
+
+	ui := state.Get("ui").(packer.Ui)
+	ssmconn := ssm.New(s.AWSSession)
+	_, err := ssmconn.TerminateSession(&ssm.TerminateSessionInput{SessionId: s.ssmSession.SessionId})
+	if err != nil {
+		msg := fmt.Sprintf("Error terminating SSM Session %q. Please terminate the session manually: %s",
```

Comment:
I think this should be auto killed

Reply:
I have to look into this a bit more as the session termination is not for the actual tunnel but for the active session within the AWS Systems Manager service. The session is created by the SDK and will get automatically terminated if there is no activity within the session for 20 minutes. The call to terminate here is to close the data connection between the Session Manager client and SSM Agent on the instance.

Reply:
ah, super nice I understand now 👍 sounds good to me :D. This is most likely already on your mind but I'd document that too maybe.